### PR TITLE
FLAT multivalue knn search implementation

### PR DESCRIPTION
**Pull request by GuyAv46**
**Describe the changes in the pull request**

This PR enables KNN search with multivalue indexing. the returned results should have each label result at most once.

**Main objects this PR modified**
1. Defined abstract max priority queue class.
2. Changed the current definition of the max priority queue to derive from the abstract class.
3. Implemented updatable max heap.
4. Made the multi-value BF use this heap.
5. Generalize the BF base flow to support both heaps.

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

**Review by alonre24 — COMMENTED** [review]
Very good! added some comments (mainly small comments and asks for clarifications). Also:
Consider adding python bindings and flow tests on random vectors in the multi index for both stability and performance tests

File: src/VecSim/algorithms/hnsw/hnsw_batch_iterator.h [hunk 1]
```diff
@@ -10,8 +10,8 @@ typedef unsigned int idType;
 using namespace std;
 using spaces::dist_func_t;
 
-using candidatesMinHeap = vecsim_stl::min_priority_queue<pair<float, idType>>;
-using candidatesMaxHeap = vecsim_stl::max_priority_queue<pair<float, idType>>;
+using candidatesMinHeap = vecsim_stl::min_priority_queue<idType, float>;
+using candidatesMaxHeap = vecsim_stl::max_priority_queue<idType, float>;
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw_batch_iterator.h:14]: I thought it more intuitive to have the value first and its priority second, but we can revert this if we disagree on it...

File: tests/unit/test_bruteforce_multi.cpp [hunk 2]
```diff
@@ -178,251 +178,257 @@ TEST_F(BruteForceMultiTest, brute_force_empty_index) {
...
+}
 
-//     float query[] = {50, 50, 50, 50};
-//     auto verify_res = [&](size_t id, float score, size_t index) { ASSERT_EQ(id, (index + 45)); };
-//     runTopKSearchTest(index, query, k, verify_res, nullptr, BY_ID);
+TEST_F(BruteForceMultiTest, search_more_then_there_is) {
+    size_t dim = 4;
+    size_t n = 5;
+    size_t perLabel = 3;
+    size_t labels = (n % perLabel) ? n / perLabel + 1 : n / perLabel;
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply tests/unit/test_bruteforce_multi.cpp:224]: Just for simple modification, if someone wants to play with them. It implies that these three values are bounded together and cannot be modified independently.

File: tests/unit/test_bruteforce_multi.cpp [hunk 3]
```diff
@@ -221,7 +221,7 @@ TEST_F(BruteForceMultiTest, search_more_then_there_is) {
     size_t dim = 4;
     size_t n = 5;
     size_t perLabel = 3;
-    size_t labels = (n % perLabel) ? n / perLabel + 1 : n / perLabel;
+    size_t n_labels = (n % perLabel) ? n / perLabel + 1 : n / perLabel;
```
Comment (alonre24) [inline tests/unit/test_bruteforce_multi.cpp:224]: `ceil(n_labels / perLabel)

**Review by alonre24 — APPROVED** [review]

File: tests/unit/test_bruteforce_multi.cpp [hunk 4]
```diff
@@ -179,251 +179,296 @@ TEST_F(BruteForceMultiTest, brute_force_empty_index) {
...
+    VecSimIndex *index = VecSimIndex_New(&params);
 
-// TEST_F(BruteForceMultiTest, brute_force_reindexing_same_vector) {
-//     size_t n = 100;
-//     size_t k = 10;
-//     size_t dim = 4;
-//     size_t initial_capacity = 200;
+    // Building the index. Each label gets 10 vectors with decreasing (by insertion order) element
+    // value, so when we search, each vector is better then the previous one. Furthermore, each
+    // label gets at leat one better vector then the previous label, and one with score equals to
```
Comment (DvirDukhan) [inline tests/unit/test_bruteforce_multi.cpp:323]:
```suggestion
    // label gets at least one better vector than the previous label and one with a score equal to
```

File: tests/unit/test_bruteforce_multi.cpp [hunk 5]
```diff
@@ -179,251 +179,296 @@ TEST_F(BruteForceMultiTest, brute_force_empty_index) {
...
+    // the best of the previous label, so the multimap holds at least two labels with the same
+    // score.
+    std::map<size_t, float> scores;
+    for (size_t i = 0; i < n; i++) {
+        float f[dim];
+        size_t el = ((n - i - 1) % n_labels) + ((n - i) / n_labels);
```
Comment (DvirDukhan) [inline tests/unit/test_bruteforce_multi.cpp:329]: Can you explain (in a comment) how for example, label 1 gets one vector with a score equal to label 2?

File: tests/unit/test_bruteforce_multi.cpp [hunk 6]
```diff
@@ -179,251 +179,296 @@ TEST_F(BruteForceMultiTest, brute_force_empty_index) {
...
+        VecSimIndex_AddVector(index, (const void *)f, i);
+    }
+    // Add more vectors under the same labels. their scores should be worst.
+    for (size_t i = 0; i < n; i++) {
+        float f[dim];
+        for (size_t j = 0; j < dim; j++) {
+            f[j] = (float)(i / 10) + n;
+        }
+        for (size_t j = 0; j < perLabel - 1; j++) {
+            VecSimIndex_AddVector(index, (const void *)f, i);
```
Comment (DvirDukhan) [inline tests/unit/test_bruteforce_multi.cpp:423]: You add here the same vector 3 times for label `i`? is that what you meant?

File: src/VecSim/utils/vecsim_stl.h [hunk 7]
```diff
@@ -23,15 +23,58 @@ class vector : public VecsimBaseObject, public std::vector<T, VecsimSTLAllocator
...
+};
+
 // max-heap
-template <typename T, typename Container = vecsim_stl::vector<T>,
-          typename Compare = std::less<typename Container::value_type>>
-using max_priority_queue = std::priority_queue<T, Container, Compare>;
+template <typename P, typename V>
+struct max_priority_queue
+    : public priority_queue_abstract<P, V>,
+      public std::priority_queue<std::pair<P, V>, vecsim_stl::vector<std::pair<P, V>>,
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/utils/vecsim_stl.h]: No. You can't pass an allocator to the priority queue itself, and we use `vecsim_stl::vector` for the container, and it uses the VecSim allocator.

File: src/VecSim/utils/vecsim_stl.h [hunk 8]
```diff
@@ -23,15 +23,44 @@ class vector : public VecsimBaseObject, public std::vector<T, VecsimSTLAllocator
...
+    inline const std::pair<P, V> top() const override { return max_pq.top(); }
+    inline size_t size() const override { return max_pq.size(); }
+};
 
 // min-heap
-template <typename T, typename Container = vecsim_stl::vector<T>,
-          typename Compare = std::greater<typename Container::value_type>>
-using min_priority_queue = std::priority_queue<T, Container, Compare>;
+template <typename P, typename V>
+using min_priority_queue = std::priority_queue<std::pair<P, V>, vecsim_stl::vector<std::pair<P, V>>,
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/utils/vecsim_stl.h:62]: Not yet. we might need it on the HNSW algorithm

File: tests/unit/test_bruteforce_multi.cpp [hunk 9]
```diff
@@ -179,251 +179,296 @@ TEST_F(BruteForceMultiTest, brute_force_empty_index) {
...
-//         for (size_t j = 0; j < dim; j++) {
-//             f[j] = (float)(i / 10); // i / 10 is in integer (take the "floor" value).
-//         }
-//         VecSimIndex_AddVector(index, (const void *)f, i);
-//     }
-//     ASSERT_EQ(VecSimIndex_IndexSize(index), n);
+    float query[] = {0, 0, 0, 0};
+    auto verify_res = [&](size_t id, float score, size_t index) { ASSERT_EQ(id, index); };
+    runTopKSearchTest(index, query, k, verify_res);
+    auto res = VecSimIndex_TopKQuery(index, query, k, nullptr, BY_SCORE);
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply tests/unit/test_bruteforce_multi.cpp:291]: Not entirely, `runTopKSearchTest` also test for the uniqueness of the results and the number of results

File: tests/unit/test_bruteforce_multi.cpp (hunk 6 above)
Comment (GuyAv46) [reply tests/unit/test_bruteforce_multi.cpp:423]: Yes. This is a test from the BF tests, and I simply want to check that it passes on multi-value as well.

**Review by DvirDukhan — APPROVED** [review]
👍

**Pull request merged**